tracing: elide tags from recording if not verbose

tbg opened the pull request:
This avoids some of the obviously problematic expensive stringifications
observed in https://github.com/cockroachdb/cockroach/issues/59379.

Release justification: low-risk performance optimization that unblocks sql stats.
Release note: None

tbg committed:
tracing: elide tags from recording if not verbose

This avoids some of the obviously problematic expensive stringifications
observed in https://github.com/cockroachdb/cockroach/issues/59379.

Release justification: low-risk performance optimization that unblocks sql stats.
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61116)
<!-- Reviewable:end -->

tbg committed:
cluster: perf optimize tracer's version feature gate

`version.IsActive` unmarshals a protobuf (and this isn't trivial to
fix), so let the tracer cache a positive result. This means that in
the common case, IsActive is not on the hot path any more.

Release justification: low-risk performance improvement
Release note: None

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61116#-:-MUNmMZoCxatEkey_sI6:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
Not very familiar with tracing yet, but an alternative might be to generate and memoize the tags on first use.

> Reviewed 5 of 5 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61116#-:-MUO3HtBEfsHDuduZzay:b-geaf7b)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61116#-:-MUO3xxW3fFbD4GGKzQf:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r+

> Not very familiar with tracing yet, but an alternative might be to generate and memoize the tags on first use.

This is difficult with the way things currently work: tags are stringified ~once per Span, so memoization would have to live outside of the tracing package. But I agree that it would be workable.
